### PR TITLE
Add stake pool mode to the wallet

### PR DESCRIPTION
**Pull request by cjepson**
A stake pool mode for the wallet has been added that only inserts
tickets paying out to an extended public key passed to the server.
The extended public key and the index of the public key to scan to
are passed as the new command line or configuration argument
"enablestakepool". Required fees for the pool are passed in the
previously added "poolfees" argument.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -88,6 +89,7 @@ type config struct {
 	PoolAddress       string  `long:"pooladdress" description:"The ticket pool address where ticket fees will go to"`
 	PoolFees          float64 `long:"poolfees" description:"The per-ticket fee mandated by the ticket pool, in coins"`
 	AddrIdxScanLen    int     `long:"addridxscanlen" description:"The width of the scan for last used addresses on wallet restore and start up (default: 750)"`
+	EnableStakePool   string  `long:"enablestakepool" description:"Enables the wallet as a stake pool and takes an extended key in the format of \"xpub...:index\" to derive cold wallet addresses to send fees to"`
```
Comment (dajohi) [inline config.go]: This configuration option needs a rename.

File: config.go [hunk 2]
```diff
@@ -90,6 +91,7 @@ type config struct {
 	PoolAddress       string  `long:"pooladdress" description:"The ticket pool address where ticket fees will go to"`
 	PoolFees          float64 `long:"poolfees" description:"The per-ticket fee mandated by the ticket pool, in coins"`
 	AddrIdxScanLen    int     `long:"addridxscanlen" description:"The width of the scan for last used addresses on wallet restore and start up (default: 750)"`
+	StakePoolFeeAddrs string  `long:"stakepoolfeeaddrs" description:"Enables the wallet as a stake pool with an extended key in the format of \"xpub...:index\" to derive cold wallet addresses to send fees to"`
```
Comment (dajohi) [inline config.go]: this takes a single value..  so addrs looks wrong..

Comment (cjepson) [inline config.go]: It's value + index of addresses to get
dpub...:idx = address(dpub..., 0), ... , address(dpub..., idx)

Comment (davecgh) [inline config.go]: If it takes multiple it should be made consistent with the other options that take multiple entries in my opinion.

```Go
StakePoolFeeAddrs []string
```

**Comment by jolan** [comment]
Works for me on testnet with a stakepool setup.

tACK

File: config.go (hunk 2 above)
Comment (cjepson) [inline config.go]: As it derives, that won't work.

Maybe PoolFeeExtPub?

**Comment by alexlyp** [comment]
tACK, looks good.  we'll need to get some more doco to along with this to make sure everyone understands the various options and ways to set it up iwth stake pools

File: wallet/chainntfns.go [hunk 3]
```diff
@@ -338,6 +342,82 @@ func (w *Wallet) handleReorganizing(oldHash *chainhash.Hash, oldHeight int64,
...
+					"out amt for commit in vout %v: %s", i, err.Error())
+			}
+			in += dcrutil.Amount(commitAmt)
+		}
+	}
+	out := dcrutil.Amount(0)
+	for i := range tx.TxOut {
+		out += dcrutil.Amount(tx.TxOut[i].Value)
+	}
+	fees := out - in
```
Comment (jrick) [inline wallet/chainntfns.go]: Shouldn't this be `in - out`?

File: wallet/chainntfns.go [hunk 4]
```diff
@@ -338,6 +342,82 @@ func (w *Wallet) handleReorganizing(oldHash *chainhash.Hash, oldHeight int64,
...
+			return false, nil
+		}
+	} else {
+		log.Warnf("Unknown pool commitment address %s for ticket %v",
+			commitAddr.EncodeAddress(), tx.TxSha())
+		return false, nil
+	}
+
+	log.Debugf("Accepted valid stake pool ticket %v commiting %v in fees",
+		tx.TxSha(), dcrutil.Amount(tx.TxOut[0].Value).ToCoin())
```
Comment (jrick) [inline wallet/chainntfns.go]: Don't use ToCoin.  This will return a float64 and is not guaranteed to be logged with the correct precision.

File: wallet/chainntfns.go [hunk 5]
```diff
@@ -397,8 +480,58 @@ func (w *Wallet) addRelevantTx(rec *wtxmgr.TxRecord,
...
+				} else {
+					// We are operating as a stake pool. The below
+					// function will ONLY add the ticket into the
+					// stake pool if it has been found within a
+					// block.
+					if block == nil {
+						break
+					}
+
+					valid, err := w.evaluateStakePoolTicket(rec, block,
```
Comment (jrick) [inline wallet/chainntfns.go]: missing a check for this error....

File: wallet/chainntfns.go [hunk 6]
```diff
@@ -419,10 +552,40 @@ func (w *Wallet) addRelevantTx(rec *wtxmgr.TxRecord,
...
+				poolTicket.HeightSpent = uint32(block.Height)
+
+				poolUser, err := w.StakeMgr.SStxAddress(&txInHash)
+				if err != nil {
+					log.Warnf("Failed to fetch stake pool user for "+
+						"ticket %v (voted ticket)", txInHash)
+				} else {
+					err = w.StakeMgr.UpdateStakePoolUserTickets(poolUser,
+						poolTicket)
+					if err != nil {
```
Comment (jrick) [inline wallet/chainntfns.go]: shouldn't this error be returned?

Comment (cjepson) [inline wallet/chainntfns.go]: No, it's an update to a metadata bucket and shouldn't result in a failure to otherwise process the vote.

File: wallet/chainntfns.go [hunk 7]
```diff
@@ -690,21 +886,24 @@ func (w *Wallet) addRelevantTx(rec *wtxmgr.TxRecord,
...
-		if err != nil {
-			log.Errorf("Cannot query transaction details for notifiation: %v", err)
-		} else {
-			w.NtfnServer.notifyUnminedTransaction(details)
-		}
-	} else {
-		details, err := w.TxStore.UniqueTxDetails(&rec.Hash, &block.Block)
-		if err != nil {
-			log.Errorf("Cannot query transaction details for notifiation: %v", err)
+	if w.NtfnServer != nil && (len(w.NtfnServer.accountClients) > 0 ||
```
Comment (jrick) [inline wallet/chainntfns.go]: why was this added? notification server should always be enabled.

File: wallet/txrules/poolfees.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2016 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+// Package txrules provides transaction rules that should be followed by
```
Comment (jrick) [inline wallet/txrules/poolfees.go]: this doc comment should be removed, it's already in another file.

File: wallet/txrules/poolfees.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/decred/dcrd/blockchain"
+	"github.com/decred/dcrd/chaincfg"
+	"github.com/decred/dcrutil"
+)
+
+const maxPoolFee = 10000.0
+
+// TestPoolFee tests to see if a pool fee is a valid percentage from 0.01% to
+// 100.00%.
+func TestPoolFee(fee float64) error {
```
Comment (jrick) [inline wallet/txrules/poolfees.go]: I'd avoid starting this with Test as it could cause confusion with a unit test.  I'd also say FeePercentage instead of Fee.

File: wallet/txrules/poolfees.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+//    ps = (f * (v+z)/(v+s)) obtained from the knowledge
+//    that the outputs of the vote are the amounts
+//    of the stake ticket plus subsidy (v+s) scaled by
+//    the proportional input of the stake pool fee
+//    f/(v+z).
+//
+// f is then adjusted for the fact that at least one subsidy reduction is
+// likely to occur before it can vote on a block.
+//
+// This all must be done with integers, so bear with the big.Int usage below.
```
Comment (jrick) [inline wallet/txrules/poolfees.go]: this comment is an implementation detail so don't put it in the doc comment.

File: wallet/txrules/poolfees.go (hunk 9 above)
Comment (davecgh) [inline wallet/txrules/poolfees.go]: `IsValidPoolFeeRate`

File: wallet/chainntfns.go (hunk 3 above)
Comment (cjepson) [inline wallet/chainntfns.go]: fixed

File: wallet/chainntfns.go (hunk 7 above)
Comment (cjepson) [inline wallet/chainntfns.go]: The calls for UniqueTxDetails are really expensive and if they can be avoided it'd be nice. I didn't think the NtfnServer for wallet were commonly used so I added the check for them.

Is

`(len(w.NtfnServer.accountClients) > 0 || len(w.NtfnServer.spentness) > 0) {`

correct for skipping the ntfn in the event there are no clients? Maybe should be another PR

Comment (jrick) [inline wallet/chainntfns.go]: even if this optimization is desired, we don't want it in this pr because it's unrelated.

File: wallet/txrules/poolfees.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// StakePoolTicketFee determines the stake pool ticket fee for a given ticket
+// from the passed percentage. Pool fee as a percentage is truncated from 0.01%
+// to 100.00%. This all must be done with integers, so bear with the big.Int
+// usage below.
+//
+// See the included doc.go of this package for more information about the
+// calculation of this fee.
+func StakePoolTicketFee(stakeDiff dcrutil.Amount, fee dcrutil.Amount,
```
Comment (jrick) [inline wallet/txrules/poolfees.go]: is fee the relay fee (looks like it based on the test vectors)? can it be renamed to relayFee?

Comment (cjepson) [inline wallet/txrules/poolfees.go]: Yeah, that'd make things more clear

File: wallet/chainntfns.go (hunk 7 above)
Comment (cjepson) [inline wallet/chainntfns.go]: OK, I'll make a separate PR
